### DISCUSSION TITLE
how to convert 1.png to 3.png

### DISCUSSION THREAD

**Discussion by linyouq** [ImageMagick/ImageMagick#4327] [Help] [answered]
![](https://s3.bmp.ovh/imgs/2021/10/dee01f1ab244d024.png)

**Comment by fmw42** [comment]
What is your ImageMagick version and platform?  Here is the commands for ImageMagick 6 on Linux or Mac OSX.

Line1: read the input
Line2: copy the input and fill with white
Line3: copy the input, convert to gray and stretch so only black and white and save as mpr and delete the copy
Line4: read the mpr, blur, negate, threshold so again only black and white, use morphology to fill small holes
Line5: (continued) read the mpr again and multiply composite it with the Line4 result
Line6: use Line5 as a mask to composite the result of Line2 over Line1 images
Line7: save the result

Input:

![1](https://user-images.githubusercontent.com/13458218/136417890-d6ecb17d-b28b-4004-8451-e954edbdf8b3.png)

```
convert 1.png \
\( -clone 0 -fill white -colorize 100 \) \
\( -clone 0 -colorspace gray -auto-level +write mpr:bw +delete \) \
\( mpr:bw -blur 0x3 -negate -threshold 0% -morphology close disk:2 \
mpr:bw -compose multiply -composite \) \
-compose over -composite \
1proc.png

```

Result:

![1proc](https://user-images.githubusercontent.com/13458218/136417974-cf6562d6-9ba8-4d3d-9806-8029726eeef3.png)

You will need to adjust the blur and threshold values to adjust the size of the white area.

Next time, please post individual images.

**Reply by linyouq** [reply]
thanks you !

i can't run you command in wondows

![](https://s3.bmp.ovh/imgs/2021/10/15286b740fe73eef.png)


this is the image
![](https://s3.bmp.ovh/imgs/2021/10/d4b52de1aaec5041.png)

I found this implementation, but it should be wrong. 
```
magick 1.png -motion-blur 0x50+45 -motion-blur 0x50+225 -motion-blur 0x50+135 -motion-blur 0x50+315 -fill 'none' -opaque black -channel RGB -threshold -1 2.png
magick 1.png -compose dst-over 2.png -composite 3.png
```
So how can I modify your command

**Comment by fmw42** [comment]
In Windows my command would be

```
convert.exe 1.png ^
( -clone 0 -fill white -colorize 100 ) ^
( -clone 0 -colorspace gray -auto-level +write mpr:bw +delete ) ^
( mpr:bw -blur 0x3 -negate -threshold 0% -morphology close disk:2 ^
mpr:bw -compose multiply -composite ) ^
-compose over -composite ^
1proc.png

```

The ^ is special character for end of line. So I have multiple lines that need to be kept. If you want to remove the ^, then make all one long line.

convert.exe 1.png ( -clone 0 -fill white -colorize 100 ) ( -clone 0 -colorspace gray -auto-level +write mpr:bw +delete ) ( mpr:bw -blur 0x3 -negate -threshold 0% -morphology close disk:2 mpr:bw -compose multiply -composite ) -compose over -composite 1proc.png

**Reply by linyouq** [reply]
The command can run, but the result is wrong.
Is it because the image has alpha channel? or my ImageMagick version

**Reply by fmw42** [reply]
What is your version?  Please always post your IM version and platform/OS when asking questions, since syntax may be different? Are you running it in a .bat file or in the CMD window? If using ImageMagick 7, then change convert.exe to magick.exe.  Post the input and output images you are using? If it is a different image or different size, then the command may need to be changed for the blur or threshold.

**Reply by linyouq** [reply]
input
![11](https://user-images.githubusercontent.com/15726523/136441671-4ce83c38-597a-43ba-bdca-266af600a2fd.png)
output
![11proc](https://user-images.githubusercontent.com/15726523/136441689-1f433977-f0f7-4d7c-bbbc-31b5cacba97a.png)

Version: ImageMagick 7.0.10-25 Q16 x64 2020-08-01
OS: windows

running it in the CMD window
`magick 11.png ( -clone 0 -fill white -colorize 100 ) ( -clone 0 -colorspace gray -auto-level +write mpr:bw +delete ) ( mpr:bw -blur 0x50 -negate -threshold 0% -morphology close disk:2 mpr:bw -compose multiply -composite ) -compose over -composite 11proc.png`

**Reply by fmw42** [reply]
Your input is completely different from the first example you showed. There is no color background. So placing white on white will not show the white contour around the text.  Also the scale is so much larger. You would have to increase the blur by a factor proportional to the scale.  

So what do you really want for the output here?

**Reply by fmw42** [reply]
You also now have an alpha channel. The original example does not.  Do you want to use the new image and make an output that has a colored background with white around the text?

**Reply by linyouq** [reply]
yes.i want to make an output that has a colored background with white around the text or other  content.


Sorry to provide the wrong example。

The correct example is
input

![1](https://user-images.githubusercontent.com/15726523/136445053-17e36e18-fe8b-40d4-9828-3e7eb730a3ac.png)
![11](https://user-images.githubusercontent.com/15726523/136445066-8ba55224-b1ac-407a-84ba-97846309a018.png)
output
![3](https://user-images.githubusercontent.com/15726523/136445104-c6c9eb37-b803-42f3-ada9-e5e7adfea32c.png)
![33](https://user-images.githubusercontent.com/15726523/136445111-2d23a6d2-8683-4abb-9c7c-e64bae76236c.png)

**Comment by fmw42** [comment, marked as answer]
Your image is way to large to process properly. So I have resize by a factor of 10% to 625 pixels on a side. Here is a different command. Adjust the morphology dilate disk:20 in proportion to the ratio dimensions of some other image compared to 625.

Input:

![1small](https://user-images.githubusercontent.com/13458218/136449182-eba59584-1d89-4c5c-bbfd-d278830b21bf.png)

magick.exe 1small.png ( -clone 0 -alpha extract -write mpr:alpha +delete ) -background skyblue -alpha background -alpha off ( -clone 0 -fill white -colorize 100 ) ( mpr:alpha -morphology dilate disk:20 ( mpr:alpha -negate ) -compose multiply -composite ) -alpha off -compose over -composite 1small_proc.png

Result:

![1small_proc](https://user-images.githubusercontent.com/13458218/136449406-aedce32b-5cca-4b01-8f17-161a007ff685.png)

**Reply by fmw42** [reply]
You can do the automatic scaling as

magick.exe 1small.png ( -clone 0 -alpha extract -write mpr:alpha +delete ) -background skyblue -alpha background -alpha off ( -clone 0 -fill white -colorize 100 ) ( mpr:alpha -morphology dilate disk:"%[fx:20*max(w,h)/625]" ( mpr:alpha -negate ) -compose multiply -composite ) -alpha off -compose over -composite 1small_proc.png

to get the same result as above

**Comment by snibgo** [comment]
"-morphology dilate" with large kernels is slow. A faster method with a similar result uses "-blur" with a threshold. For example (Windows BAT syntax):
```
%IMG7%magick ^
  %1 ^
  ( -clone 0 -alpha extract -write mpr:alpha +delete ) ^
  -background skyblue -alpha background -alpha off ^
  ( -clone 0 -fill white -colorize 100 ) ^
  ( mpr:alpha -blur 0x%%[fx:20*max(w,h)/625] ^
    ( mpr:alpha -negate ) ^
    -compose multiply -composite ^
    -threshold 1%% ^
+write x0.png ^
  ) ^
  -alpha off -compose over -composite ^
  s2.png
```
For the large "lake and back" image, this takes 28 seconds, compared to 13 minutes using "-morphology dilate"
![s2](https://user-images.githubusercontent.com/33812368/136472410-d0715bb9-d28c-4a0d-b593-2546e927b3af.png)
.

**Reply by fmw42** [reply]
Yes, that is what I used originally. But I changed to the morphology method since it had fewer and more understandable arguments for this novice user. I did warn that his images were very large and would not process we'll with the morphology method. But thanks for the additional information and approach in Windows syntax.

**Reply by linyouq** [reply]
Version: ImageMagick 7.0.10-25 Q16 x64 2020-08-01
OS: windows

Through these commands, I achieved that result. Although I don't know why it works.

```
magick 1.png -resize 6250x6250 11.png
magick 11.png -channel RGB -threshold -1 -fill "rgb(38,39,46)" -opaque white 111.png
magick 111.png -blur 0x60% -negate -threshold 0% 1111.png
magick 1111.png -colorspace RGB -compose over 11.png -composite 11111.png
```
the input and out: https://1drv.ms/f/s!AsAOb3941MOHhEti_dHjzyLj5APn

Finally, how to convert these commands into one sentenc

**really appreciate your help and instruction**
